### ISSUE TITLE
OWASP Core Rule Set (CRS) installed by default

### ISSUE THREAD

**Issue by AnBuKu** [debops/debops#75]
In order to improve security of DebOps web hosting plan, e.g.  CRS support of [Apache server](https://github.com/debops/ansible-apache) and [Ngnix server](https://github.com/debops/ansible-nginx) DebOps roles, implementation of [ModSecurity Core Rule Set (CRS) ](https://coreruleset.org) default installation into DebOps seems appropriate.

How this benefit can be achieved - either by a new, separate DebOps role or by adding CRS to  DebOps roles Apache server and Ngnix server - is not clear at the moment and needs technical clarification.

OWASP Core Rule Set (CRS) on Debian tracker:
[Debian modsecurity-crs](https://tracker.debian.org/pkg/modsecurity-crs)

Contacts for questions:
- Debian package maintainer:
Alberto Gonzalez Iniesta, email: agi@inittab.org
- ModSecurity Core Rule Set main contributor:
Christian Folini on [Github](https://github.com/dune73) and on [OWASP](https://www.owasp.org/index.php/User:Dune73), email: christian.folini@netnea.com

**Comment by drybjed** [comment]
At the moment, Debian includes a support for `mod_security` using a [`libapache2-mod-security2`](https://packages.debian.org/stretch/libapache2-mod-security2) package, which is meant to be used with Apache 2, but there's currently no similar package for Nginx - to have support it, users would need to recompile the `nginx` source package and provide it through a local APT repository. Since from Debian Stretch, the `nginx` Debian packages support dynamically loadable modules, I would be interested to know, if the `mod_security` team plans to include support for nginx with CRS in time that the current Debian Testind release is frozen in preparation for the Debian Buster release.

in DebOps, Apache 2 is currently only used with the ownCloud role. Personally, I haven't used it yet, but I think that each application requires different set of rules to work correctly, for example ownCloud has [`owncloud/mod_security](https://github.com/owncloud/mod_security) GitHub repository which contains the rules for that application. And there seems to be a lot of it.

I'm not sure if cramming such ruleset directly into the `debops.owncloud` role is a good idea. Perhaps there could be a separate, optional role, that enables the `mod_security` support in Apache 2 and manages the rulesets on the hosts.

If there are people interested in working on support for this feature, let me know. I've added this issue to the Debian Buster milestone, so there's still at least a year to go, lots of time to make this right.

**Comment by dune73** [comment]
Author of the ModSecurity Handbook and CRS project co-lead here.

A good overview over what CRS is, what it does and how it works can be found at [https://lwn.net/Articles/709693/](url).

ModSecurity 2.9.x is not entirely stable on NGINX. Hence no package.

ModSecurity 3.0 will be stable on NGINX but development is slow, the state is beta and it's not quite feature complete as of this writing. Funny note: NGINX Plus has binaries with ModSecurity 3.0 which they confirm are stable and advanced enough to run CRS on.

CRS 2.x had a ton of false positives. Like 2-5 FPs per request without tuning.

CRS 3.0 (release November 2016) had the release goal "Remove FPs from the default installation". This was achieved to a very wide extent. We have at least hundreds of thousands of installations, but received only 22 FP releated features for the default installation in the year since the release.

The owncloud/mod_security repository linked above points to CRS 2.x.

We support rule exclusion packages that can be enabled within our config files to tune away known FPs up to "paranoia level 2". The default installation is at PL1, at PL2 you have to expect some FPs, hence the exclusion package. For DebOps I would clearly aim for a baseline security setup with a default installation: Hence PL1. It is also possible to raise the anomaly scoring threshold, thus letting some attacks pass, but also allowing some FPs to go unblocked. This lowers the security benefit of course, but it can be a good entry point when you are unsure you trust the rule set on your production setup.

TL;DR; CRS 3.0 will work out of the box for most installations. There will be FPs with the default installation, but they will be rather rare. It is important to point users to documentation that helps them
getting rid of FPs like the tutorials at [netnea.com](url).

**Comment by drybjed** [comment]
Hello @dune73, thanks for the explanation.

So I guess there's no need to focus on `nginx` at present, let's wait for the support to stabilize. Apache 2 might be an interesting choice, however it isn't widely supported in DebOps yet. I haven't used the role too much either, it was written by @ypid, so I'll let him worry about it for now, if he's interested. :-)

**Comment by dune73** [comment]
You are most welcome, @drybjed. Glad to be of assistance.

**Comment by ypid** [comment]
> it was written by @ypid, so I'll let him worry about it for now, if he's interested. :-)

I am :-)

Ref: [apache__security_module_enabled](https://docs.debops.org/en/latest/ansible/roles/ansible-apache/docs/defaults.html#envvar-apache__security_module_enabled)

Note that I currently don’t use Apache in production. I am only responsible for Nginx at the moment. The Apache role is in a good state I would say but is looking for a security conscious dev to continue what I started :)

I agree with @drybjed and will look into better hardening for Nginx when appropriate.

**Comment by AnBuKu** [comment]
Looks to me like CRS is a hot potato :-D

**Comment by drybjed** [comment]
@AnBuKu for me, not really - I'm currently focused on finishing various tests for the DebOps monorepo, for example testing as much of the roles as possible using a combination of GitLab, Vagrant, LXC and libvirt. After that, I'll probably will be busy cleaning up the Ansible and Python syntax in the repository. When that's done, DebOps will have its first stable release. After that, I'll start bringing more roles and updating existing ones.

It's hard to tell about the timeframe of all of this for the moment. I think that the test suite + documentation will be ready to go around January-February 2018, we will see.

**Comment by AnBuKu** [comment]
@drybjed no stress, was just pulling the legs of you guys. :-)

I appreciate all the efforts of you guys. Even me knows, how much you deliver. Thus, no hard feelings at all. :-)
